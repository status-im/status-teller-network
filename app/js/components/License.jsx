import React from 'react';
import {Card, CardHeader, CardBody, CardTitle, Button, Alert, ButtonGroup} from 'reactstrap';
import PropTypes from 'prop-types';

const BuyLicense = (props) => (
  <Button onClick={props.buyLicense}>Buy License</Button>
);
BuyLicense.propTypes = {
  buyLicense: PropTypes.func
};

const IsLicenseOwner = () => <p>You already own a license</p>;

const Buttons = ({rate}) => {
  const buttons = [];
  for (let i = 1; i <= 5; i++) {
    buttons.push(<Button key={'rating-' + i} onClick={() => rate(i)}>{i}</Button>);
  }
  return buttons;
};

const License = (props) => (
  <Card className="mt-2">
    <CardHeader>
      <CardTitle>License</CardTitle>
    </CardHeader>
    <CardBody>
      {props.error && <Alert color="danger">{this.props.error}</Alert>}
      {props.isLicenseOwner ? <IsLicenseOwner/> : <BuyLicense buyLicense={props.buyLicense}/>}
      <p>Rating: {props.userRating ? props.userRating : '-'}</p>

      <form>
        <label className="mr-3">Rate the transaction:</label>
        <ButtonGroup>
          <Buttons rate={props.rate}/>
        </ButtonGroup>
      </form>
    </CardBody>
  </Card>
);

License.propTypes = {
  error: PropTypes.string,
  isLicenseOwner: PropTypes.bool,
  userRating: PropTypes.number,
  rate: PropTypes.func,
  buyLicense: PropTypes.func
};

export default License;
