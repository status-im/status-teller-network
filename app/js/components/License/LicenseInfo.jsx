import React, {Component} from 'react';
import {Row, Col, Card, CardBody} from 'reactstrap';
import PropTypes from 'prop-types';
import {withNamespaces} from 'react-i18next';

class LicenseInfo extends Component {
  render() {
    const t = this.props.t;
    return (
      <div className="mt-5 mb-5">
        <Row>
          <Col xs={12}>
            <h2 className="text-center">{t('sellerLicenseInfo.title')}</h2>
          </Col>
        </Row>

        <Card color="primary" className="mt-5 py-4">
          <CardBody className="text-center">
            {t('sellerLicenseInfo.fee')}
          </CardBody>
        </Card>

        <Card color="primary" className="mt-5 py-4">
          <CardBody className="text-center">
            {t('sellerLicenseInfo.stake', {price: this.props.price})}
          </CardBody>
        </Card>
      </div>
    );
  }
}

LicenseInfo.propTypes = {
  t: PropTypes.func,
  price: PropTypes.string
};

export default withNamespaces()(LicenseInfo);
