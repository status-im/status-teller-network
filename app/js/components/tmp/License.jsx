import React from 'react';
import {Card, CardHeader, CardBody, CardTitle, Button} from 'reactstrap';
import PropTypes from 'prop-types';
import { withNamespaces } from 'react-i18next';
import TransactionResults from "./TransactionResults";

const BuyLicense = (props) => (
  <Button onClick={props.buyLicense} disabled={props.disabled}>{props.t('license.buy')}</Button>
);

BuyLicense.propTypes = {
  t: PropTypes.func,
  buyLicense: PropTypes.func,
  disabled: PropTypes.bool
};

const IsLicenseOwner = (props) => <p>{props.t('license.alreadyOwner')}</p>;

IsLicenseOwner.propTypes = {
  t: PropTypes.func
};

const License = (props) => (
  <Card className="mt-2">
    <CardHeader>
      <CardTitle>{props.t('license.title')}</CardTitle>
    </CardHeader>
    <CardBody>
      <TransactionResults txHash={props.txHash} loading={props.loading} error={props.error}
                          loadingText={props.t('license.buying')} errorText={"Error: "}/>
      {props.isLicenseOwner ? <IsLicenseOwner t={props.t}/> : <BuyLicense buyLicense={props.buyLicense} t={props.t} disabled={props.loading}/>}
      {props.isLicenseOwner &&
        <p>{props.t('license.rating')} {props.userRating ? props.userRating : props.t('license.noRating')}</p>
      }
    </CardBody>
  </Card>
);

License.propTypes = {
  t: PropTypes.func,
  error: PropTypes.string,
  txHash: PropTypes.string,
  isLicenseOwner: PropTypes.bool,
  loading: PropTypes.bool,
  userRating: PropTypes.number,
  buyLicense: PropTypes.func
};

export default withNamespaces()(License);
