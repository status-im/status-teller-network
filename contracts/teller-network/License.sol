/* solium-disable security/no-block-members */
/* solium-disable security/no-inline-assembly */

pragma solidity  ^0.5.8;

import "../common/Ownable.sol";
import "../token/ERC20Token.sol";
import "../token/ApproveAndCallFallBack.sol";

/**
* @title License
* @dev Contract for buying a license
*/
contract License is Ownable, ApproveAndCallFallBack {
    uint256 public price;
    string private constant LICENSE_ALREADY_BOUGHT = "License already bought";
    string private constant UNSUCCESSFUL_TOKEN_TRANSFER = "Unsuccessful token transfer";

    ERC20Token token;

    struct LicenseDetails {
        uint price;
        uint creationTime;
    }

    mapping(address => LicenseDetails) public licenseDetails;

    address[] public licenseOwners;

    mapping(address => uint) public idxLicenseOwners;

    event Bought(address buyer, uint256 price);
    event PriceChanged(uint256 _price);

    constructor(address payable _tokenAddress, uint256 _price) public {
        price = _price;
        token = ERC20Token(_tokenAddress);
    }

    /**
    * @dev Check if the address already owns a license
    * @param _address The address to check
    * @return bool
    */
    function isLicenseOwner(address _address) public view returns (bool) {
        return licenseDetails[_address].price != 0 && licenseDetails[_address].creationTime != 0;
    }

    /**
    * @dev Buy a license
    * @notice Requires value to be equal to the price of the license.
    *         The msg.sender must not already own a license.
    */
    function buy() public {
        buyFrom(msg.sender);
    }

    /**
    * @dev Buy a license
    * @notice Requires value to be equal to the price of the license.
    *         The _owner must not already own a license.
    */
    function buyFrom(address _owner) private {
        require(licenseDetails[_owner].creationTime == 0, LICENSE_ALREADY_BOUGHT);
        require(token.allowance(_owner, address(this)) >= price, "Allowance not set for this contract to expected price");
        require(token.transferFrom(_owner, address(this), price), UNSUCCESSFUL_TOKEN_TRANSFER);

        licenseDetails[_owner] = LicenseDetails({
            price: price,
            creationTime: block.timestamp
        });

        uint idx = licenseOwners.push(_owner);
        idxLicenseOwners[_owner] = idx;

        emit Bought(_owner, price);
    }

    /**
    * @dev Set the price
    * @param _price The new price of the license
    * @notice Only the owner of the contract can perform this action
    */
    function setPrice(uint256 _price) public onlyOwner {
        price = _price;
        emit PriceChanged(_price);
    }

    /**
     * @dev Get number of license owners
     * @return uint
     */
    function getNumLicenseOwners() public view returns (uint256) {
        return licenseOwners.length;
    }

    /**
     * @notice Withdraw not reserved tokens
     * @param _token Address of ERC20 withdrawing excess, or address(0) if want ETH.
     * @param _beneficiary Address to send the funds.
     **/
    function withdrawExcessBalance(address _token, address payable _beneficiary) external onlyOwner {
        require(_beneficiary != address(0), "Cannot burn token");
        if (_token == address(0)) {
            _beneficiary.transfer(address(this).balance);
        } else {
            ERC20Token excessToken = ERC20Token(_token);
            uint256 amount = excessToken.balanceOf(address(this));
            excessToken.transfer(_beneficiary, amount);
        }
    }


    /**
     * @notice Support for "approveAndCall". Callable only by `token()`.
     * @param _from Who approved.
     * @param _amount Amount being approved, need to be equal `price()`.
     * @param _token Token being approved, need to be equal `token()`.
     * @param _data Abi encoded data with selector of `register(bytes32,address,bytes32,bytes32)`.
     */
    function receiveApproval(address _from, uint256 _amount, address _token, bytes memory _data) public {
        require(_amount == price, "Wrong value");
        require(_token == address(token), "Wrong token");
        require(_token == address(msg.sender), "Wrong call");
        require(_data.length == 4, "Wrong data length");

        bytes4 sig = abiDecodeRegister(_data);

        require(
            sig == bytes4(0xa6f2ae3a), //bytes4(keccak256("buy()"))
            "Wrong method selector"
        );

        buyFrom(_from);
    }


    /**
     * @dev Decodes abi encoded data with selector for "buy()".
     * @param _data Abi encoded data.
     * @return Decoded registry call.
     */
    function abiDecodeRegister(bytes memory _data) public pure returns(bytes4 sig) {
        assembly {
            sig := mload(add(_data, add(0x20, 0)))
        }
    }
}
