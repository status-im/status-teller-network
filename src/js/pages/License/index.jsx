import React, {Component} from 'react';
import {withRouter} from "react-router-dom";
import PropTypes from 'prop-types';
import {connect} from "react-redux";

import license from "../../features/license";
import network from "../../features/network";

import Info from './components/Info';
import BuyButton from './components/BuyButton';
import Balance from './components/Balance';
import Loading from '../../components/Loading';
import ErrorInformation from '../../components/ErrorInformation';

const LICENSE_TOKEN_SYMBOL = 'SNT';

class License extends Component {
  componentDidMount() {
    if (this.props.isLicenseOwner) {
      return this.props.history.push('/sell');
    }
    this.pollBalanceInterval = null;

    this.props.checkLicenseOwner();
    this.props.loadLicensePrice();
    this.props.updateBalance(LICENSE_TOKEN_SYMBOL);
  }

  checkBalance() {
    if (this.props.sntToken && this.props.sntToken.balance) {
      if (this.enoughBalance()) {
        if (this.pollBalanceInterval) {
          clearInterval(this.pollBalanceInterval);
        }
      } else {
        if (!this.pollBalanceInterval) {
          this.pollBalanceInterval = setInterval(() => {
            this.props.updateBalance(LICENSE_TOKEN_SYMBOL);
          }, 2000);
        }
      }
    }
  }

  componentDidUpdate() {
    if (this.props.isLicenseOwner) {
      return this.props.history.push('/sell');
    }
    this.checkBalance();
  }

  componentWillUnmount() {
    if (this.pollBalanceInterval) {
      clearInterval(this.pollBalanceInterval);
    }
  }

  buyLicense = () => {
    this.setState({isBuying: true});
    this.props.buyLicense();
  };

  enoughBalance() {
    return this.props.sntToken && parseInt(this.props.sntToken.balance, 10) >= parseInt(this.props.licensePrice, 10);
  }

  render() {
    if (!this.props.sntToken) {
      return <ErrorInformation sntTokenError retry={this.buyLicense}/>;
    }
    if (this.props.error) {
      return <ErrorInformation transaction retry={this.buyLicense} message={this.props.error} cancel={this.props.cancelBuyLicense}/>;
    }

    if (this.props.isLoading) {
      return <Loading mining txHash={this.props.txHash}/>;
    }

    return (
      <React.Fragment>
        <Info price={this.props.licensePrice} />
        <div className="mt-5">
          <Balance value={this.props.sntToken.balance} disabled={!this.enoughBalance()}/>
        </div>
        <BuyButton onClick={this.buyLicense} disabled={!this.enoughBalance()}/>
      </React.Fragment>
    );
  }
}

License.propTypes = {
  history: PropTypes.object,
  wizard: PropTypes.object,
  checkLicenseOwner: PropTypes.func,
  buyLicense: PropTypes.func,
  isLicenseOwner: PropTypes.bool,
  isLoading: PropTypes.bool,
  txHash: PropTypes.string,
  error: PropTypes.string,
  sntToken: PropTypes.object,
  licensePrice: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number
  ]),
  loadLicensePrice: PropTypes.func,
  updateBalance: PropTypes.func,
  cancelBuyLicense: PropTypes.func
};

const mapStateToProps = state => {
  return {
    isLicenseOwner: license.selectors.isLicenseOwner(state),
    isLoading: license.selectors.isLoading(state),
    txHash: license.selectors.txHash(state),
    error: license.selectors.error(state),
    sntToken: network.selectors.getTokenBySymbol(state, LICENSE_TOKEN_SYMBOL),
    licensePrice: license.selectors.getLicensePrice(state)
  };
};

export default connect(
  mapStateToProps,
  {
    buyLicense: license.actions.buyLicense,
    cancelBuyLicense: license.actions.cancelBuyLicense,
    checkLicenseOwner: license.actions.checkLicenseOwner,
    loadLicensePrice: license.actions.loadPrice,
    updateBalance: network.actions.updateBalance
  }
)(withRouter(License));
