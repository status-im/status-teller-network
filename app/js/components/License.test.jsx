import React from "react";
import { shallow } from "enzyme";

import License from "./License";

describe('License', () => {
  it('should render correctly', () => {
    const component = shallow(<License />);
  
    expect(component).toMatchSnapshot();
  });

  it('should render when being license owner', () => {
    const component = shallow(<License isLicenseOwner />);
  
    expect(component).toMatchSnapshot();
  });
});
