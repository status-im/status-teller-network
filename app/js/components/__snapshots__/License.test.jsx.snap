// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`License should render correctly 1`] = `
<Card
  className="mt-2"
  tag="div"
>
  <CardHeader
    tag="div"
  >
    <CardTitle
      tag="h5"
    />
  </CardHeader>
  <CardBody
    tag="div"
  >
    <BuyLicense
      t={[Function]}
    />
  </CardBody>
</Card>
`;

exports[`License should render when being license owner 1`] = `
<Card
  className="mt-2"
  tag="div"
>
  <CardHeader
    tag="div"
  >
    <CardTitle
      tag="h5"
    />
  </CardHeader>
  <CardBody
    tag="div"
  >
    <IsLicenseOwner
      t={[Function]}
    />
    <p>
       
    </p>
  </CardBody>
</Card>
`;
